packages/imt.sol/contracts/internal/InternalLazyIMT.sol

```diff
@@ -157,6 +157,74 @@ library InternalLazyIMT {
         return _root(self, numberOfLeaves, depth);
     }
 
+    function _merkleProofElements(
+        LazyIMTData storage self,
+        uint40 index,
+        uint8 depth
+    ) internal view returns (uint256[] memory) {
+        uint40 numberOfLeaves = self.numberOfLeaves;
+        require(index < numberOfLeaves, "LazyIMT: leaf must exist");
+        require(depth > 0, "LazyIMT: depth must be > 0");
+        require(depth <= MAX_DEPTH, "LazyIMT: depth must be < MAX_DEPTH");
+
+        uint256[] memory proof = new uint256[](depth);
+        uint256 levelCount = self.numberOfLeaves;
+
+        for (uint8 i = 0; i < depth; ) {
+            // Left leaf
+            if (index & 1 == 0) {
+                bool outsideLimits = ((index + 1) << i) % self.numberOfLeaves != 0;
+                if ((index + 1) < levelCount) {
+                    proof[i] = self.elements[_indexForElement(i, index + 1)];
+                } else if (((index + 1) == levelCount) && (outsideLimits)) {
+                    proof[i] = _root(self, numberOfLeaves, i);
```

Comment:
This function is calling into the `_root` function which is building the tree up to this level. This results in _many_ duplicate hashes being calculated, something like `O(N^2)` vs `O(N)`.

I think this function can be entirely removed and the elements of the `levels` array in `_root` can be returned directly. One approach to this could be splitting the logic from `_root` into a function that returns the levels, then in `_root` just returning the top level item, and in `_merkleProofElements` returning the levels.

Something like below should work and minimize duplicate code. The types/logic will need to be fixed but the structure should work.

```ts
function _root(LazyIMTData storage self, uint40 numberOfLeaves, uint8 depth) internal view returns (uint256) {
         // this should always short circuit if self.numberOfLeaves == 0
        if (numberOfLeaves == 0) return _defaultZero(depth);
        uint256 memory levels = _levels(self, numberOfLeaves, depth, numberOfLeaves - 1);
        return levels[depth];
}

function _merkleProofElements(LazyIMTData storage self, uint40 numberOfLeaves, uint8 depth, uint40 index) internal view returns (uint256) {
        uint256 memory levels = _levels(self, numberOfLeaves, depth, index);
        return levels;
}

function _levels(LazyIMTData storage self, uint40 numberOfLeaves, uint8 depth, uint40 index) internal view returns (uint256[] memory) {
        require(depth <= MAX_DEPTH, "LazyIMT: depth must be <= MAX_DEPTH");
       
        // accept as argument instead of hardcoding
        // uint40 index = numberOfLeaves - 1;
        uint256[MAX_DEPTH + 1] memory levels;
        if (index & 1 == 0) {
            levels[0] = self.elements[_indexForElement(0, index)];
        } else {
            levels[0] = _defaultZero(0);
        }
        for (uint8 i = 0; i < depth; ) {
            if (index & 1 == 0) {
                levels[i + 1] = PoseidonT3.hash([levels[i], _defaultZero(i)]);
            } else {
                uint256 levelCount = (numberOfLeaves) >> (i + 1);
                if (levelCount > index >> 1) {
                    uint256 parent = self.elements[_indexForElement(i + 1, index >> 1)];
                    levels[i + 1] = parent;
                } else {
                    uint256 sibling = self.elements[_indexForElement(i, index - 1)];
                    levels[i + 1] = PoseidonT3.hash([sibling, levels[i]]);
                }
            }
            unchecked {
                index >>= 1;
                i++;
            }
        }
        return levels;
}

```

Reply:
@vimwitch Acknowledge the problem and love the idea of reusing `_levels` for both `root` and `_merkleProofElements` but not sure this works as expected. Aren't the hashes different?

I mean, when going bottom->up, `root` calculates the hashes using the right or left element (either _defaultZero(x) or an existing leaf) but in `_merkleProofElements`, we are interested in the opposite leaf to construct the proof.

Am I missing something?

Reply:
Yes i think you're right, the logic isn't exactly the same. I'd advise copying the logic from the `_root` function and changing it a bit. The merkle path function should be calling Poseidon directly. It might be possible to tweak such an internal `_levels` function to work for both, but not strictly necessary.

Also one other thing i forgot to mention, instead of calling it `MerkleTreeIndexes` i think `MerkleTreeIndices` is more correct.

Reply:
> Also one other thing i forgot to mention, instead of calling it MerkleTreeIndexes i think MerkleTreeIndices is more correct.
right! done: 19c0b431ff44a779a46591d8d5516083214a8319

> Yes i think you're right, the logic isn't exactly the same. I'd advise copying the logic from the _root function and changing it a bit. The merkle path function should be calling Poseidon directly. It might be possible to tweak such an internal _levels function to work for both, but not strictly necessary.

@vimwitch Would this optimization be blocking for this PR? Not sure my other commitments would allow me to allocate time for this, or at least in a while.

Reply:
@vimwitch Note that for each tree I'm getting the merkle proof of all leafs, so that's around 700 calls to the `_merkleProofElements` in total.

Feel free to apply the changes, really curious about the results. Thanks!

Reply:
Yes true, the test case is large.

I opened a pr alrevuelta/zk-kit#1. The approach i described mostly works. Each time merkle elements are loaded a merkle proof for the latest leaf is built to construct any subtrees that haven't been built yet. It should reduce the number of poseidon hashes that are required on average to get the elements, though the performance difference isn't notable in the current test cases.

Let me know what you think.

Reply:
Seems indeed way more efficient. Thanks!

Reply:
Thanks for the help. This should solve this comment: https://github.com/alrevuelta/zk-kit/pull/1